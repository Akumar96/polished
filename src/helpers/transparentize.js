// @flow
import { isRgb, isRgba, isHsl, isHsla, isHex } from '../internalHelpers/_isValidColor'
import hexToRgb from '../internalHelpers/_hexToRgb'
import hslToRgb from '../internalHelpers/_hslToRgb'

const getNewTransparency = (currentTransparency: number, percentage: number) : number => (
  currentTransparency * (1 - percentage)
)

const getColorValuesInBrackets = (color: string) : string => (
  color.slice(color.indexOf('(') + 1, color.indexOf(')'))
)

const getValuesTillLastComma = (color: string) : string => (
  color.slice(0, color.lastIndexOf(','))
)

const transparentizeRgb = (rgb: string, percentage: number) : string => (
  `rgba(${getColorValuesInBrackets(rgb)}, ${getNewTransparency(1, percentage)})`
)

const getLastColorValue = (color: string) : number => (
  Number(
    color.slice(
      color.lastIndexOf(',') + 1,
      color.indexOf(')'),
    ),
  )
)

function transparentize(color: string, percentage: number) {
  if (percentage >= 0 && percentage <= 1) {
    if (isHex(color)) {
      return transparentizeRgb(`${hexToRgb(color)})`, percentage)
    } else if (isRgb(color)) {
      return transparentizeRgb(color, percentage)
    } else if (isRgba(color)) {
      const currentTransparency = getLastColorValue(color)
      const rgb = getValuesTillLastComma(getColorValuesInBrackets(color))
      const newTransparency = getNewTransparency(currentTransparency, percentage)

      return `rgba(${rgb}, ${newTransparency})`
    } else if (isHsl(color)) {
      const newTransparency = getNewTransparency(1, percentage)
      const values = getColorValuesInBrackets(color).split(',').map(num => parseInt(num, 10) / 100)
      const rgbValues = hslToRgb(...values)

      return `rgba(${rgbValues}, ${newTransparency})`
    } else if (isHsla(color)) {
      const hslValues = getValuesTillLastComma(getColorValuesInBrackets(color))
      const currentTransparency = getLastColorValue(color)
      const newTransparency = getNewTransparency(currentTransparency, percentage)
      const values = hslValues.split(',').map(num => parseInt(num, 10) / 100)
      const rgbValues = hslToRgb(...values)

      return `rgba(${rgbValues}, ${newTransparency})`
    } else {
      throw new Error('Invalid color')
    }
  } else {
    throw new Error('Invalid percentage, only values from 0 to 1 are accepted')
  }
}

export default transparentize
