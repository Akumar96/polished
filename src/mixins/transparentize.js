// @flow
import { isRgb, isRgba, isHsl, isHsla, isHex } from '../internalHelpers/_isValidColor'
import hslToRgb from '../internalHelpers/_hslToRgb'
import hexToRgb from '../internalHelpers/_hexToRgb'

const getNewTransparency = (currentTransparency: number, percentage: number) : number => (
  currentTransparency * (1 - (percentage / 100))
)

const getColorValuesInBrackets = (color: string) : string => (
  color.slice(color.indexOf('(') + 1, color.indexOf(')'))
)

const getValuesTillLastComma = (color: string) : string => (
  color.slice(0, color.lastIndexOf(','))
)

const transparentizeRgb = (rgb: string, percentage: number) : string => (
  `rgba(${getColorValuesInBrackets(rgb)}, ${getNewTransparency(1, percentage)})`
)

const getLastColorValue = (color: string) : number => (
  Number(
    color.slice(
      color.lastIndexOf(',') + 1,
      color.indexOf(')'),
    ),
  )
)

function transparentize(color: string, percentage : number) {
  if (isHex(color)) {
    return transparentizeRgb(`${hexToRgb(color)})`, percentage)
  } else if (isRgb(color)) {
    return transparentizeRgb(color, percentage)
  } else if (isRgba(color)) {
    const currentTransparency = getLastColorValue(color)
    const rgb = getValuesTillLastComma(getColorValuesInBrackets(color))
    const newTransparency = getNewTransparency(currentTransparency, percentage)

    return `rgba(${rgb}, ${newTransparency})`
  } else if (isHsl(color)) {
    const hslValues = getColorValuesInBrackets(color)
    const rgb = hslToRgb(
      ...(hslValues.split(', ').map(val => parseInt(val, 10))),
    )

    return transparentizeRgb(rgb, percentage)
  } else if (isHsla(color)) {
    const hslValues = getValuesTillLastComma(getColorValuesInBrackets(color))
    const currentTransparency = getLastColorValue(color)
    const rgb = hslToRgb(
      ...(hslValues.split(', ').map(val => parseInt(val, 10))),
    )

    const newTransparency = getNewTransparency(currentTransparency, percentage)
    return `rgba(${rgb}, ${newTransparency})`
  } else {
    throw new Error('Invalid color')
  }
}

export default transparentize
