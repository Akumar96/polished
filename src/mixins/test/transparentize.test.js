// @flow
import transparentize from '../transparentize'

describe('transparentize', () => {
  it('it should correctly convert hex #fff with 20% transparency to rgba', () => {
    expect(transparentize('#fff', 20)).toMatchSnapshot()
  })

  it('it should correctly convert rgb(255, 0, 255) with 10% transparency to rgba', () => {
    expect(transparentize('rgb(255, 0, 255)', 10)).toMatchSnapshot()
  })

  it('it should correctly decrease current alpha of rgba(255, 0, 0, 1) with 10% transparency to rgba', () => {
    expect(transparentize('rgba(255, 0, 0, 1)', 10)).toMatchSnapshot()
  })

  it('it should correctly decrease current alpha of rgba(255, 0, 0, .5) by 50%', () => {
    expect(transparentize('rgba(255, 0, 0, .5)', 50)).toMatchSnapshot()
  })

  it('it should correctly decrease current alpha of rgba(255, 0, 0, .5) by 50%', () => {
    expect(transparentize('rgba(255, 0, 0, .5)', 50)).toMatchSnapshot()
  })

  it('it should correctly convert hsl(0, 0%, 100%) with 20% transparency to rgba', () => {
    expect(transparentize('hsl(0, 0%, 100%)', 20)).toMatchSnapshot()
  })

  it('it should correctly convert hsla(0, 0%, 100%, .8) with 50% transparency to rgba', () => {
    expect(transparentize('hsla(0, 0%, 100%, .8)', 50)).toMatchSnapshot()
  })

  it('it should throw an error when enter an invalid color', () => {
    expect(() => { transparentize('not a color', 50) }).toThrow(new Error('Invalid color'))
  })
})
