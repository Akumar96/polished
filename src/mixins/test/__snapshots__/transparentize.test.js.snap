exports[`transparentize it should correctly convert hex #fff with 20% transparency to rgba 1`] = `"rgba(255, 255, 255, 0.8)"`;

exports[`transparentize it should correctly convert hsl(0, 0%, 100%) with 20% transparency to rgba 1`] = `"hsla(0, 0%, 100%, 0.8)"`;

exports[`transparentize it should correctly convert hsla(0, 0%, 100%, .8) with 50% transparency to rgba 1`] = `"hsla(0, 0%, 100%, 0.4)"`;

exports[`transparentize it should correctly convert rgb(255, 0, 255) with 10% transparency to rgba 1`] = `"rgba(255, 0, 255, 0.9)"`;

exports[`transparentize it should correctly decrease current alpha of rgba(255, 0, 0, .5) by 50% 1`] = `"rgba(255, 0, 0, 0.25)"`;

exports[`transparentize it should correctly decrease current alpha of rgba(255, 0, 0, .5) by 50% 2`] = `"rgba(255, 0, 0, 0.25)"`;

exports[`transparentize it should correctly decrease current alpha of rgba(255, 0, 0, 1) with 10% transparency to rgba 1`] = `"rgba(255, 0, 0, 0.9)"`;
